docs/blog/posts/chain-of-density.md

```diff
@@ -0,0 +1,467 @@
+    be returned when using the `instructions.distil` annotation
+
+!!! warning "Rate Limiting"
+
+    We recommend running this script on a small subset of the dataset first to test you've got everything configured nicely.
+    Don't forget to add in rate limiting error handling with `tenacity` and set the `OPENAI_API_KEY` shell environment variable
+    before running any subsequent commands
+
+### Creating Fine-Tuning Jobs
+
+Once we run this script, we'll have a new file called `generated.jsonl` in our local repository. Now all that's left is to run the command below to start fine-tuning your first model!
+
+```sh
+instructor jobs create-from-file generated.jsonl
+```
+
+??? notes "Finetuning Reference"
+
+    Checking out our [Finetuning CLI](/instructor/cli/finetune/) to learn about other hyperparameters that you can tune to improve your model's performance.
+
+Once the job is complete, all we need to do is to then change the annotation in the function call to `distil_summarization` in our original file above to start using our new model.
+
+```py
+@instructions.distil(model='gpt-3.5-turbo:finetuned-123', mode="dispatch") #(1)!
+def distil_summarization(text: str) -> GeneratedSummary:
+    summary_chain: List[str] = summarize_article(text)
+    return GeneratedSummary(summary=summary_chain[-1])
+```
+
+1. Don't forget to replace this with your new model id. OpenAI identifies fine tuned models with an id of
+   ft:gpt-3.5-turbo-0613:personal::<id> under their Fine-tuning tab on their dashboard
+
+With that, you've now got your own fine-tuned model ready to go and serve data in production. We've seen how Instructor can make your life easier, from fine-tuning to distillation.
+
+## Results and Benchmarks
+
+We fine-tuned a total of 3 different models, giving each 20, 50 and 76 samples respectively to see if more data improved the models. We then compared the output of these fine tuned models to GPT-4 and GPT-3 summaries that were generated using chain-of-density methods.
+
+We'll be comparing these models in three main ways
+
+- Entity Density : This is entities per token, the higher the better for density.
+- Latency : Time to last token generated in seconds
+- Costs : How much does the entire experiment cost
+
+We used a total of 20 articles as a validation set which our fine tuned models had not seen before. This was the overall performance that we observed.
+
+| Model               | Mean Latency (s) | Mean Entity Count | Mean Entity Density | Tokens |
+| ------------------- | ---------------- | ----------------- | ------------------- | ------ |
+| GPT-4 (COD)         | 49.5             | 11.3              | 0.138               | 81.65  |
+| GPT-3 (COD)         | 145.94           | 11.05             | 0.105               | 105.7  |
+| 3.5 Finetuned (20)  | 2.25             | 14.7              | 0.154               | 95.45  |
```

Comment:
Comparing finetuning on 20 summaries to the rows below, it seems that finetuning only on 20 summaries had the highest absolute entity count and density? Hmm, why is that? Might want to add a few hypotheses even if you don't have the answers.

Reply:
Added it in a new branch! But my hypotheses are that

1. Model might not be benefitting from the higher number of examples due to the epochs ( 20, 50 and 76 are all trained with 4 epochs which is the default number provided by OpenAI)
2. Larger variety of examples might cause the model to optimize for different objectives - not just for token density. The COD summarization method tends to produce more abstract summaries with each rewrite, so it might be learning other metrics under the hood to optimize for